### ISSUE TITLE
Snyk:  vm2 Sandbox Bypass Vulnerability | SNYK-JS-VM2-5537100

### ISSUE THREAD

**Issue by tony-snoop** [snowflakedb/snowflake-connector-nodejs#512]
# Overview

[vm2](https://github.com/patriksimek/vm2#readme) is a sandbox that can run untrusted code with whitelisted Node's built-in modules.

Affected versions of this package are vulnerable to Sandbox Bypass by abusing an unexpected creation of a host object based on the maliciously crafted specification of Proxy. Exploiting this vulnerability allows an attacker to gain remote code execution rights on the host running the sandbox via the Function constructor.

# Introduced through:

snowflake-sdk@1.6.22 › urllib@2.38.0 › proxy-agent@5.0.0 › pac-proxy-agent@5.0.0 › pac-resolver@5.0.1 › degenerator@3.0.3 › vm2@3.9.17

Could the `urllib` package be updated to a newer version, to fix the downstream vulnerability identified with the `vm2` package?

# Links:

https://security.snyk.io/vuln/SNYK-JS-VM2-5537100
https://www.cve.org/CVERecord?id=CVE-2023-32314

**Comment by tony-snoop** [comment]
https://github.com/snowflakedb/snowflake-connector-nodejs/blob/master/package.json

```
{
  "name": "snowflake-sdk",
  "version": "1.6.22",
  "description": "Node.js driver for Snowflake",
  "dependencies": {
    "@azure/storage-blob": "^12.11.0",
    "@techteamer/ocsp": "1.0.0",
    "agent-base": "^6.0.2",
    "asn1.js-rfc2560": "^5.0.0",
    "asn1.js-rfc5280": "^3.0.0",
	"async": "^3.2.3",
    "aws-sdk": "^2.878.0",
    "axios": "^0.27.2",
    "better-eval": "^1.3.0",
    "big-integer": "^1.6.43",
    "bignumber.js": "^2.4.0",
    "binascii": "0.0.2",
    "bn.js": "^5.2.1",
    "browser-request": "^0.3.3",
    "debug": "^3.2.6",
    "expand-tilde": "^2.0.2",
    "extend": "^3.0.2",
    "fast-xml-parser": "^4.1.3",
    "generic-pool": "^3.8.2",
    "@google-cloud/storage": "^6.9.3",
    "glob": "^7.1.6",
    "https-proxy-agent": "^5.0.1",
    "jsonwebtoken": "^9.0.0",
    "mime-types": "^2.1.29",
    "mkdirp": "^1.0.3",
    "mock-require": "^3.0.3",
    "moment": "^2.29.4",
    "moment-timezone": "^0.5.15",
    "open": "^7.3.1",
    "python-struct": "^1.1.3",
    "simple-lru-cache": "^0.0.2",
    "string-similarity": "^4.0.4",
    "test-console": "^2.0.0",
    "tmp": "^0.2.1",
    "urllib": "^2.38.0",                              <----- can this be upgraded to a newer version?
    "uuid": "^3.3.2",
    "winston": "^3.1.0"
  },
  "devDependencies": {
    "qs": "^6.10.3",
    "mocha": "^10.1.0"
  },
  "overrides": {
    "@azure/storage-blob": {
      "node-fetch": "^3.2.10"
    }
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/snowflakedb/snowflake-connector-nodejs"
  },
  "scripts": {
    "test": "mocha -timeout 180000 --recursive --full-trace test/unit/**/*.js test/unit/*.js",
    "test:integration": "mocha -timeout 180000 --recursive --full-trace test/integration/**/*.js test/integration/*.js",
    "test:single": "mocha -timeout 180000 --full-trace",
    "test:system": "mocha -timeout 180000 --recursive --full-trace system_test/*.js",
    "test:unit": "mocha -timeout 180000 --recursive --full-trace test/unit/**/*.js test/unit/*.js",
    "test:ci": "mocha -timeout 180000 --recursive --full-trace test/**/*.js system_test/*.js"
  },
  "author": {
    "name": "Snowflake Computing, Inc.",
    "email": "support@snowflake.com",
    "url": "https://www.snowflake.com/"
  },
  "license": "Apache-2.0"
}
```

**Comment by sfc-gh-dszmolka** [comment]
hi and thank you for raising this issue. with a fresh install of `snowflake-sdk` (version 1.6.22)
```
# npm ls vm2
node@ /node
`-- snowflake-sdk@1.6.22
  `-- urllib@2.40.0
    `-- proxy-agent@5.0.0
      `-- pac-proxy-agent@5.0.0
        `-- pac-resolver@5.0.1
          `-- degenerator@3.0.4
            `-- vm2@3.9.19   <--- per CVE-2023-32314 this is not affected
```

since `degenerator` [depends on vm2 ^3.9.17](https://github.com/TooTallNate/proxy-agents/blob/main/packages/degenerator/package.json#L30), it allows for fixed versions (`3.9.18+`) of `vm2` to be installed, which is exactly what is happening here.

can you please try reinstalling `snowflake-sdk` and rescan?

**Comment by tony-snoop** [comment]
Hi @sfc-gh-dszmolka ,

I've had a look at the dependency chain found in the package-lock.json file and you're correct - it's referencing vm2@3.9.19.
An upgrade to the latest version of snowflake-sdk@1.6.22, and a rescan has corrected the snyk issue!

Apologies for the inconvenience - and many thanks for the quick investigation and reply.

```

    "snowflake-sdk": {
      "version": "1.6.22",
      "resolved": "https://snoop-471456929700.d.codeartifact.eu-west-1.amazonaws.com/npm/snoop-shd-shared/snowflake-sdk/-/snowflake-sdk-1.6.22.tgz",
      "integrity": "sha512-EAGcO0Fx4a91JK79BSlgMzLxFNaRPP5T6SewvtCPfylcOF4wFyn4ih9HrpyRqgbUlJCUitsTYEr2G7VTb25DKw==",
      "requires": {
        "@azure/storage-blob": "^12.11.0",
        "@google-cloud/storage": "^6.9.3",
        "@techteamer/ocsp": "1.0.0",
        "agent-base": "^6.0.2",
        "asn1.js-rfc2560": "^5.0.0",
        "asn1.js-rfc5280": "^3.0.0",
        "async": "^3.2.3",
        "aws-sdk": "^2.878.0",
        "axios": "^0.27.2",
        "better-eval": "^1.3.0",
        "big-integer": "^1.6.43",
        "bignumber.js": "^2.4.0",
        "binascii": "0.0.2",
        "bn.js": "^5.2.1",
        "browser-request": "^0.3.3",
        "debug": "^3.2.6",
        "expand-tilde": "^2.0.2",
        "extend": "^3.0.2",
        "fast-xml-parser": "^4.1.3",
        "generic-pool": "^3.8.2",
        "glob": "^7.1.6",
        "https-proxy-agent": "^5.0.1",
        "jsonwebtoken": "^9.0.0",
        "mime-types": "^2.1.29",
        "mkdirp": "^1.0.3",
        "mock-require": "^3.0.3",
        "moment": "^2.29.4",
        "moment-timezone": "^0.5.15",
        "open": "^7.3.1",
        "python-struct": "^1.1.3",
        "simple-lru-cache": "^0.0.2",
        "string-similarity": "^4.0.4",
        "test-console": "^2.0.0",
        "tmp": "^0.2.1",
        "urllib": "^2.38.0",                     <--- 
        "uuid": "^3.3.2",
        "winston": "^3.1.0"
      },
...

    "urllib": {
      "version": "2.40.0",
      "resolved": "https://snoop-471456929700.d.codeartifact.eu-west-1.amazonaws.com/npm/snoop-shd-shared/urllib/-/urllib-2.40.0.tgz",
      "integrity": "sha512-XDZjoijtzsbkXTXgM+A/sJM002nwoYsc46YOYr6MNH2jUUw1nCBf2ywT1WaPsVEWJX4Yr+9isGmYj4+yofFn9g==",
      "requires": {
        "any-promise": "^1.3.0",
        "content-type": "^1.0.2",
        "debug": "^2.6.9",
        "default-user-agent": "^1.0.0",
        "digest-header": "^1.0.0",
        "ee-first": "~1.1.1",
        "formstream": "^1.1.0",
        "humanize-ms": "^1.2.0",
        "iconv-lite": "^0.4.15",
        "ip": "^1.1.5",
        "proxy-agent": "^5.0.0",               <---
        "pump": "^3.0.0",
        "qs": "^6.4.0",
        "statuses": "^1.3.1",
        "utility": "^1.16.1"
      },
...
    "proxy-agent": {
      "version": "5.0.0",
      "resolved": "https://snoop-471456929700.d.codeartifact.eu-west-1.amazonaws.com/npm/snoop-shd-shared/proxy-agent/-/proxy-agent-5.0.0.tgz",
      "integrity": "sha512-gkH7BkvLVkSfX9Dk27W6TyNOWWZWRilRfk1XxGNWOYJ2TuedAv1yFpCaU9QSBmBe716XOTNpYNOzhysyw8xn7g==",
      "requires": {
        "agent-base": "^6.0.0",
        "debug": "4",
        "http-proxy-agent": "^4.0.0",
        "https-proxy-agent": "^5.0.0",
        "lru-cache": "^5.1.1",
        "pac-proxy-agent": "^5.0.0",                  <--- 
        "proxy-from-env": "^1.0.0",
        "socks-proxy-agent": "^5.0.0"
      },	  
...
    "pac-proxy-agent": {
      "version": "5.0.0",
      "resolved": "https://snoop-471456929700.d.codeartifact.eu-west-1.amazonaws.com/npm/snoop-shd-shared/pac-proxy-agent/-/pac-proxy-agent-5.0.0.tgz",
      "integrity": "sha512-CcFG3ZtnxO8McDigozwE3AqAw15zDvGH+OjXO4kzf7IkEKkQ4gxQ+3sdF50WmhQ4P/bVusXcqNE2S3XrNURwzQ==",
      "requires": {
        "@tootallnate/once": "1",
        "agent-base": "6",
        "debug": "4",
        "get-uri": "3",
        "http-proxy-agent": "^4.0.1",
        "https-proxy-agent": "5",
        "pac-resolver": "^5.0.0",
        "raw-body": "^2.2.0",
        "socks-proxy-agent": "5"
      },
	  

    "pac-proxy-agent": {
      "version": "5.0.0",
      "resolved": "https://snoop-471456929700.d.codeartifact.eu-west-1.amazonaws.com/npm/snoop-shd-shared/pac-proxy-agent/-/pac-proxy-agent-5.0.0.tgz",
      "integrity": "sha512-CcFG3ZtnxO8McDigozwE3AqAw15zDvGH+OjXO4kzf7IkEKkQ4gxQ+3sdF50WmhQ4P/bVusXcqNE2S3XrNURwzQ==",
      "requires": {
        "@tootallnate/once": "1",
        "agent-base": "6",
        "debug": "4",
        "get-uri": "3",
        "http-proxy-agent": "^4.0.1",
        "https-proxy-agent": "5",
        "pac-resolver": "^5.0.0",       <--
        "raw-body": "^2.2.0",
        "socks-proxy-agent": "5"
      },
...
    "pac-resolver": {
      "version": "5.0.1",
      "resolved": "https://snoop-471456929700.d.codeartifact.eu-west-1.amazonaws.com/npm/snoop-shd-shared/pac-resolver/-/pac-resolver-5.0.1.tgz",
      "integrity": "sha512-cy7u00ko2KVgBAjuhevqpPeHIkCIqPe1v24cydhWjmeuzaBfmUWFCZJ1iAh5TuVzVZoUzXIW7K8sMYOZ84uZ9Q==",
      "requires": {
        "degenerator": "^3.0.2",      <--- 
        "ip": "^1.1.5",
        "netmask": "^2.0.2"
      }
    },
...
    "degenerator": {
      "version": "3.0.4",
      "resolved": "https://snoop-471456929700.d.codeartifact.eu-west-1.amazonaws.com/npm/snoop-shd-shared/degenerator/-/degenerator-3.0.4.tgz",
      "integrity": "sha512-Z66uPeBfHZAHVmue3HPfyKu2Q0rC2cRxbTOsvmU/po5fvvcx27W4mIu9n0PUlQih4oUYvcG1BsbtVv8x7KDOSw==",
      "requires": {
        "ast-types": "^0.13.2",
        "escodegen": "^1.8.1",
        "esprima": "^4.0.0",
        "vm2": "^3.9.17"          <-- 
      }
...
    "vm2": {
      "version": "3.9.19",              <-- happy days!
      "resolved": "https://snoop-471456929700.d.codeartifact.eu-west-1.amazonaws.com/npm/snoop-shd-shared/vm2/-/vm2-3.9.19.tgz",
      "integrity": "sha512-J637XF0DHDMV57R6JyVsTak7nIL8gy5KH4r1HiwWLf/4GBbb5MKL5y7LpmF4A8E2nR6XmzpmMFQ7V7ppPTmUQg==",
      "requires": {
        "acorn": "^8.7.0",
        "acorn-walk": "^8.2.0"
      }
```
